src/transformers/models/audioflamingo3/modular_audioflamingo3.py

```diff
@@ -0,0 +1,512 @@
+
+        batch_size = len(final_embedding)
+        device = final_embedding[0].device
+        hidden_size = final_embedding[0].shape[1]
+        max_length = max(final_embedding[k].shape[0] for k in range(batch_size))
+        attention_mask = torch.ones((batch_size, max_length), dtype=torch.bool, device=device)
+
+        inputs_batched, labels_batched = [], []
+        for k in range(batch_size):
+            size_pk = max_length - final_embedding[k].shape[0]
+            inputs_pk = torch.zeros((size_pk, hidden_size), dtype=final_embedding[k].dtype, device=device)
+            labels_pk = (
+                torch.full((size_pk,), self.config.ignore_index, dtype=final_labels[k].dtype, device=device)
+                if final_labels is not None
+                else None
+            )
+
+            if self.padding_side == "right":
+                attention_mask[k, final_embedding[k].shape[0] :] = False
+                inputs_pk = torch.cat([final_embedding[k], inputs_pk], dim=0)
+                if labels_pk is not None:
+                    labels_pk = torch.cat([final_labels[k], labels_pk], dim=0)
+            else:
+                attention_mask[k, : -final_embedding[k].shape[0]] = False
+                inputs_pk = torch.cat([inputs_pk, final_embedding[k]], dim=0)
+                if labels_pk is not None:
+                    labels_pk = torch.cat([labels_pk, final_labels[k]], dim=0)
+
+            inputs_batched.append(inputs_pk)
+            if labels_pk is not None:
+                labels_batched.append(labels_pk)
+
+        inputs_embeds = torch.stack(inputs_batched, dim=0)
+        labels = torch.stack(labels_batched, dim=0) if labels_batched else None
+        return inputs_embeds, labels, attention_mask
+
+    @torch.inference_mode()
+    def generate(
+        self,
+        input_ids: Optional[torch.FloatTensor] = None,
+        audio_features: Optional[torch.Tensor] = None,  # (N, M, T_mel)
+        attention_mask: Optional[torch.LongTensor] = None,
+        audio_feature_masks: Optional[torch.Tensor] = None,  # (N, T_mel)
+        audio_features_mask: Optional[torch.Tensor] = None,  # (N, S_audio+1)
+        **generation_kwargs,
+    ) -> torch.LongTensor:
+        # Compute audio embeddings upfront
+        audio_embeds = None
+        if audio_features is not None:
+            audio_embeds = self._compute_audio_embeds(audio_features, audio_feature_masks)  # (N, S_audio+1, D)
+
+        # Merge text and audio embeddings
+        inputs_embeds, _, attn = self._merge_input_ids_with_audio_features(
+            input_ids=input_ids,
+            labels=None,
+            attention_mask=attention_mask,
+            audio_embeds=audio_embeds,
+            audio_features_mask=audio_features_mask,
+        )
+        return self.llm.generate(inputs_embeds=inputs_embeds, attention_mask=attn, **generation_kwargs)
+
+    @property
+    def default_generation_config(self) -> GenerationConfig:
```

Comment:
This is not something we normally have.

I saw that in your most [recent example](https://github.com/huggingface/transformers/pull/40290#issuecomment-3269211209), you're using this to pass the token IDs.

This is something that can be handled if you overwrite  `prepare_inputs_for_generation` as mentioned above.

Moreover, like [Dia](https://github.com/huggingface/transformers/blob/8a091cc07c9808196e3154ab8bd6252d919f929b/src/transformers/models/dia/configuration_dia.py#L340-L342) and [Llava](https://github.com/huggingface/transformers/blob/91be12bdc6d312a2d7a44174971f025c4023a91f/src/transformers/models/llava/configuration_llava.py#L87), let's have the token IDs in the configuration.